### PR TITLE
Feature/keep ui state of block items

### PR DESCRIPTION
**Pull request by planarvoid**
This PR mainly fixes keeping UI state (selected tab on Comments/Followers or expanded state of Tags & categories). This was a pretty complicated issue. In the end I'm introducing an UiState object that keeps the state of the Block and gets updated automatically. I think this is a clean solution (and architecture). The View propagates an action (for example a tab switch) to the VM and the VM creates a different model for the Block containing different data. This solution removes the logic from the view holders. I've tried to iterate on this and it's still not super easy to use. 

As part of this PR I'm also fixing the animation (and diff utils propagation of changes) when a nested recycler view changes.

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/BlockListAdapter.kt [hunk 1]
```diff
@@ -46,8 +51,19 @@ import org.wordpress.android.util.image.ImageManager
 class BlockListAdapter(val imageManager: ImageManager) : Adapter<BlockListItemViewHolder>() {
     private var items: List<BlockListItem> = listOf()
     fun update(newItems: List<BlockListItem>) {
-        items = newItems
-        notifyDataSetChanged()
+        if (newItems.size >= items.size) {
```
Comment (malinajirka) [inline WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/BlockListAdapter.kt]: We may consider adding a comment here. Wdyt?

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/BlockListDiffCallback.kt [hunk 2]
```diff
@@ -0,0 +1,23 @@
+package org.wordpress.android.ui.stats.refresh
+
+import android.support.v7.util.DiffUtil.Callback
+import org.wordpress.android.ui.stats.refresh.lists.sections.BlockListItem
+
+class BlockListDiffCallback(
```
Comment (malinajirka) [inline WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/BlockListDiffCallback.kt]: This class seems to be unused.

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/StatsBlockDiffCallback.kt [hunk 3]
```diff
@@ -23,4 +22,13 @@ class StatsBlockDiffCallback(
     override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
         return oldList[oldItemPosition] == newList[newItemPosition]
     }
+
+    override fun getChangePayload(oldItemPosition: Int, newItemPosition: Int): Any? {
+        val newItem = newList[newItemPosition]
+        val oldItem = oldList[oldItemPosition]
+        if (oldItem is BlockList && newItem is BlockList) {
```
Comment (malinajirka) [inline WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/StatsBlockDiffCallback.kt]: I'd suggest explaining here why we use the changePayload method. Wdyt?

**Review by malinajirka — COMMENTED** [review]
Thanks @planarvoid ! This solution is way more readable in my opinion, I really like it;)! GJ

I've left a few in-code comments. Last thing I mentioned but you probably forgot to fix is the way you handle `areItemsTheSame` in DiffCallback. It should return true only when the two items represent the same object -> it may result in a weird animation if you return true when the two objects represent a same type but not the same object. For example if you'd have Titles at positions 1 and 5, the diffutil may run an animation moving 1 -> 5 and 5 -> 1.

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/BaseStatsUseCase.kt [hunk 4]
```diff
@@ -60,23 +98,76 @@ abstract class BaseStatsUseCase(
...
 
-    protected fun createFailedItem(@StringRes failingType: Int, message: String): Error {
-        return Error(type, failingType, message)
+    /**
+     * Transforms given model and ui state into the UI model
+     * @param model domain model coming from FluxC
+     * @param nullableUiState contains UI specific data
+     * @return a list of block list items
+     */
+    protected abstract fun buildModel(model: Model, nullableUiState: UiState?): List<BlockListItem>
```
[the remark this answers is not in the document]
Comment (planarvoid) [reply WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/BaseStatsUseCase.kt]: Sounds good. I've renamed everything except the FluxC models

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/insights/usecases/CommentsUseCase.kt [hunk 5]
```diff
@@ -30,39 +28,50 @@ class CommentsUseCase
 @Inject constructor(
     @Named(UI_THREAD) private val mainDispatcher: CoroutineDispatcher,
     private val insightsStore: InsightsStore
-) : BaseStatsUseCase(COMMENTS, mainDispatcher) {
-    override suspend fun fetchRemoteData(site: SiteModel, forced: Boolean): StatsBlock? {
+) : StatefulUseCase<CommentsModel, Int>(COMMENTS, mainDispatcher, 0) {
```
[the remark this answers is not in the document]
Comment (planarvoid) [reply WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/insights/usecases/CommentsUseCase.kt]: nice trick 👍

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/insights/usecases/FollowersUseCase.kt [hunk 6]
```diff
@@ -36,18 +34,20 @@ class FollowersUseCase
...
+    override suspend fun loadCachedData(site: SiteModel) {
         val wpComFollowers = insightsStore.getWpComFollowers(site, PAGE_SIZE)
         val emailFollowers = insightsStore.getEmailFollowers(site, PAGE_SIZE)
-        return if (wpComFollowers != null && emailFollowers != null) loadFollowers(
-                site,
-                wpComFollowers,
-                emailFollowers
-        ) else null
+        if (wpComFollowers != null && emailFollowers != null) {
+            onModel(wpComFollowers to emailFollowers)
```
[the remark this answers is not in the document]
Comment (planarvoid) [reply WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/insights/usecases/FollowersUseCase.kt]: I've added a case to remove the block when the model is missing - the same behaviour as we get in other use cases

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/insights/usecases/FollowersUseCase.kt [hunk 7]
```diff
@@ -56,40 +56,46 @@ class FollowersUseCase
...
-        return when {
-            error != null -> createFailedItem(
-                    string.stats_view_followers,
-                    error.message ?: error.type.name
-            )
-            wpComModel != null && emailModel != null -> loadFollowers(site, wpComModel, emailModel)
-            else -> null
+        when {
+            error != null -> onError(error.message ?: error.type.name)
+            wpComModel != null && emailModel != null -> onModel(wpComModel to emailModel)
```
[the remark this answers is not in the document]
Comment (planarvoid) [reply WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/insights/usecases/FollowersUseCase.kt]: answered in the previous item

**Review by malinajirka — COMMENTED** [review]
Thanks @planarvoid. The code looks good to me! (I've left just a few nitpicks comments).

However, the ViewMore button isn't working as expected.
1. View More works just once
- open stats
- click on View More
- go back
- click on View More (nothing happens)

Other issues which are probably not related to this PR
(2.) The text in the Latest Post Summary gets cut off
screenshot_20181130-142423


(3.) I'm not sure if this is an issue or not
- Select for example "By posts & Pages" tab in Comments
- Click on ViewMore
- Notice the "By Authors" tab is selected... I'd expect the "By post & Pages" tab to be selected

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/BlockDiffCallback.kt [hunk 8]
```diff
@@ -65,12 +65,10 @@ class BlockDiffCallback(
...
         val oldItem = oldList[oldItemPosition]
-        if (newItem is ExpandableItem && oldItem is ExpandableItem) {
-            if (!oldItem.isExpanded && newItem.isExpanded) {
-                return EXPAND_ITEM
-            } else if (oldItem.isExpanded && !newItem.isExpanded) {
-                return COLLAPSE_ITEM
-            }
+        if (newItem is ExpandableItem && oldItem is ExpandableItem &&
+                ((!oldItem.isExpanded && newItem.isExpanded) ||
+                        (oldItem.isExpanded && !newItem.isExpanded))) {
```
[the remark this answers is not in the document]
Comment (planarvoid) [reply WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/BlockDiffCallback.kt]: of course 🤦‍♂️

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/BlockDiffCallback.kt [hunk 9]
```diff
@@ -24,8 +23,9 @@ class BlockDiffCallback(
     private val oldList: List<BlockListItem>,
     private val newList: List<BlockListItem>
 ) : Callback() {
-    enum class ExpandPayload {
-        EXPAND_ITEM, COLLAPSE_ITEM
+
+    enum class Payload {
```
[the remark this answers is not in the document]
Comment (planarvoid) [reply WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/BlockDiffCallback.kt]: Renamed to `BlockListPayload`

**Comment by planarvoid** [comment]
@malinajirka thanks for the review!
1) I've fixed the issue
2) I'll have one last PR for design fixes and fix everything there
3) This is still the old code. I don't think it's worth fixing in the first version and we'll rewrite the `View more` activities in v2

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/insights/usecases/FollowersUseCase.kt [hunk 10]
```diff
@@ -56,40 +61,52 @@ class FollowersUseCase
...
+            else -> {
```
Comment (malinajirka) [inline WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/insights/usecases/FollowersUseCase.kt]: Nitpick: We might want to remove the brackets.

**Review by malinajirka — COMMENTED** [review]
Thanks @planarvoid! 

The initial (after clearing app data) loading animation seems broken.
https://i.gyazo.com/f7153962ab418f32b0a49d400bb2689d.gif


I've left just a couple of nitpick comments. It looks good apart from the loading animation:).

**Comment by planarvoid** [comment]
@malinajirka everything should be fixed now

**Comment by malinajirka** [comment]
Thanks @planarvoid ! The initial loading animations looks good, unfortunately the non-initial animation seems broken - empty view is displayed for a bit.

https://i.gyazo.com/00465e782abd054e37d3a53a6e492c85.gif

**Comment by malinajirka** [comment]
It looks awesome now, thanks!!! ;)

I'll merge it as soon as Travis is done.

**Pull request merged**